### ISSUE TITLE
How to add new plugins in the fluentd Output

### ISSUE THREAD

**Issue by henrikrexed** [fluent/fluent-operator#254]
### Is your feature request related to a problem? Please describe.

it seems that adding new plugins requires to 
- modifiy the image of fluentd /fluentbit by installing the extra plugins
- modifier the various CRDs to allow our user to configure the new plugin

This constraint will slow the adoption , because lots of customers are using community  plugins

### Describe the solution you'd like

Provide a guide helping our community in building custom images.

To avoid modifying the various CRDs, we should have a way of adding "custom" Filter, output, parser steps.
modifying

### Describe alternatives you've considered

For the moment my only workaround is to use the operator to manage fluentbit, and then deploy fluentd as a daemonset to be able to use the custom plugins.

**Comment by zhu733756** [comment]
Thanks for the great ideas.

> For the moment my only workaround is to use the operator to manage fluentbit, and then deploy fluentd as a daemonset to be able to use the custom plugins.

Our design is to deploy fluentbit as a daemonset to collect logs, sink to the storage like es, loki,etc. Fluentd acts as a statefulset workload to receieve the logs from fluentbit or apps, also could sink to the storages.

**Comment by benjaminhuo** [comment]
@henrikrexed I agree with the idea of `Provide a guide helping our community in building custom images.`

Regarding the second point `to avoid modifying the various CRDs, we should have a way of adding "custom" Filter, output, parser steps`, there're two methods to support various plugins:
- add plugin settings via annotation
- add plugin settings as the current method

The annotation method is flexible but there is no validation for it.
The second is not that flexible, but there is validation for it.

We choose the second method because of the validation it provides and some fluentbit plugins are added by community contributed.

For fluentd, we'll consider the annotation method to allow more flexibility.

And we don't recommend using Fluentd as a daemonset because it's fluentbit's role（acting as a logging agent）.

Fluentd receives logs from fluentbit  or applications through network and do more advanced process.

**Comment by petergvizd** [comment]
Hello @benjaminhuo 

we would like to use fluent-operator with fluent-plugin-grok-parser, so I would like to ask, what is meant by "add plugin settings as the current method"? Do you mean by this, to add support for plugin to code, CRDs and image or something else?

**Comment by wenchajun** [comment]
> Hello @benjaminhuo
> 
> we would like to use fluent-operator with fluent-plugin-grok-parser, so I would like to ask, what is meant by "add plugin settings as the current method"? Do you mean by this, to add support for plugin to code, CRDs and image or something else?

Hi, you can refer to these use cases
https://github.com/fluent/fluent-operator/tree/master/apis/fluentbit/v1alpha2/plugins

**Comment by benjaminhuo** [comment]
There're some discussions about adding custom plugins in https://github.com/fluent/fluent-operator/issues/301#issuecomment-1152362843
 We'll see if that is a feasible approach.

**Comment by joshuabaird** [comment]
This issue is being closed due to inactivity.  Please re-open a new issue if it is still relevant.
